### ISSUE TITLE
Upgrade API for mobile apps

### ISSUE THREAD

**Issue by acadet** [rails-school/school#164]
Here are the routes we need for the mobile apps, all in JSON format. The interface can be [found there](https://github.com/rails-school/tiramisu/blob/develop/app/src/main/java/org/railsschool/tiramisu/models/bll/remote/IRailsSchoolAPI.java).

Routes can be changed but content must be similar. Also, some of them already exist, I tried to match existing configuration.

* `GET lessons/future/slugs`: Returns a list of incoming lessons' slugs only.
* `GET lessons/:slug`: Returns a lesson matching provided slug with list of attendees
* `GET lessons/upcoming`: Returns upcoming lesson

===

* `GET users/:id`: Returns user matching provided id
* `POST users/check-credentials`: System provides login + bcrypt password. It expects a permanent token as result. This token will be use when toggling attendance to lesson or checking if user is attending.
 
===

* `GET venues/:id`: Returns venue matching provided id.

===

* `PUT attendances`: System provides lesson's slug, username, user's token and attendance value (boolean). Then, website updates attendance for provided user and lesson.
* `GET attendances/:lessong_slug/:username/:user_token`: Returns true if user is attending to lesson

===

To sum up, the log in system is the following:

1. User provides username and password in mobile app.
2. Mobile app forwards them to server.
3. If correct, server returns a permanent authentication token. Otherwise, it returns a 401 status.
4. Mobile app sends username and token anytime an update relative to a user is done.

**Comment by acadet** [comment]
@donaldali suggested to use username/password instead of a token. In a way, that makes sense for if user changes his password, mobile app will keep going whereas it should not.

What is your opinion @gkop?

**Comment by gkop** [comment]
I like Donald's idea for obtaining a remember_me cookie. Then you can save the remember_me cookie to the client, and use it for subsequent authentications. I believe the remember_me tokens are invalidated automatically on password change, but you might double check.

**Comment by acadet** [comment]
Um. I don't know how to get a cookie in Android. I mean I can get any kind of data object but I guess there is no default Cookie class or something similar.

I was thinking about sending username/encrypted password anytime needed, as HTTP headers.

**Comment by gkop** [comment]
Cookies are just a special case of HTTP headers, so if you can read/write headers, you can accomplish what I describe.

**Comment by acadet** [comment]
I can, indeed. So sending this cookie will be enough?

**Comment by gkop** [comment]
Until it expires, yes.

**Comment by donaldali** [comment]
shouldn't it be possible to send the user's credentials to the new_user_session_path (/users/sign_in) with the remember_me flag and devise will set the cookie.

**Comment by acadet** [comment]
Yep. Give me the layout my POST args should match.

Ditto for toggling attendance and is attending routes.

**Comment by gkop** [comment]
Exactly @donaldali .

**Comment by donaldali** [comment]
For authentication, the POST request to /users/sign_in should submit a `params` hash/object that contains at least
```ruby
{ user: { email: "<user_email>",
          password: "<user_password>",
          remember_me: "1"
        }
}
```
but there might be authenticity_token issues if your form was not generated by Rails.

The other routes should be trivial once the authentication one works. If the remember_me doesn't work, then we might have to use an authentication_token in the users table and integrate that with devise.

**Comment by acadet** [comment]
Perfect. So am I going to receive a token in result or a cookie?

So let me know what format you expect for `isAttending` and `toggleAttendance`. And how you expect the token (http header, cookie etc.).

**Comment by donaldali** [comment]
Rememberable uses a cookie (http://www.rubydoc.info/github/plataformatec/devise/Devise/Models/Rememberable).

For toggleAttendance, attending a lesson is equivalent to RSVPing for the lesson which can be done with a POST request to `/rsvp/:id`; not attending a lesson would be unRSVPing for the lesson, which would be a POST request to `/rsvp/:id/:delete` (:id in both cases is the id of the lesson to RSVP or unRSVP and :delete should be set to "delete").
isAttending can be done with a GET request to `/attending_lesson/:lesson_slug` or `/attending_lesson/:lesson_id` or anything else we want to call it--any of which I can make.

All the routes above rely on the server knowing the current_user, which would depend on the authentication with remember_me working (or having to do it with the `authentication_token`).

**Comment by acadet** [comment]
I have just matched your authentication system. Once new website published, I will test my application.

Please let me know when it's done!

**Comment by donaldali** [comment]
This branch (https://github.com/donaldali/school/tree/upgrade-api) contains API endpoints for `/l/future/slugs`, `/l/upcoming`, and `/attending_lesson/:lesson_slug`; the last one requires the user to be authenticated.

**Comment by acadet** [comment]
Alright, so some edits you need to apply to your code @donaldali:

1. Avoid JSON requests to be revoked. Would be easier in your application controller but you can do it manually everywhere there are POST/PUT/DELETE requests I need.

```ruby
skip_before_filter :verify_authenticity_token, if: :json_request?

def json_request?
  request.format.json?
end
```

2. Enable JSON Devise support. You need `respond_to :html, :json` in your application controller + edit the `navigational_formats` in Devise's configuration.

3. Respond to JSON format in attendance controller. There is no JSON response to rsvp method at that point (status 200 is ok, I need to extra outcome).

4. Upcoming lesson route should only returns a single lesson. You return all future lessons.

5. You override `as_json` method in user model. However, `as_json` expects an optional dictionary as argument. Here, you have removed it, what produces errors when an argument is provided to the system.
